plugins/woocommerce/includes/class-wc-order-item.php

```diff
@@ -444,4 +446,102 @@ public function offsetGet( $offset ) {
 
 		return null;
 	}
+
+	/**
+	 * Indicates if the current order item has an associated Cost of Goods Sold value.
+	 *
+	 * Derived classes representing line items that have a COGS value
+	 * should override this method to return "true" and also the 'calculate_cogs_value_core' method.
+	 *
+	 * @since 9.5.0
+	 *
+	 * @return bool True if this line item has an associated Cost of Goods Sold value.
+	 */
+	public function has_cogs(): bool {
+		return false;
+	}
+
+	/**
+	 * Calculate the Cost of Goods Sold value and set it as the actual value for this line item.
+	 *
+	 * @since 9.5.0
+	 *
+	 * @return bool True if the value has been calculated successfully (and set as the actual value), false otherwise (and the value hasn't changed).
+	 * @throws Exception The class doesn't implement its own version of calculate_cogs_value_core. Derived classes are expected to override that method when has_cogs returns true.
+	 */
+	public function calculate_cogs_value(): bool {
+		if ( ! $this->has_cogs() || ! $this->cogs_is_enabled( __METHOD__ ) ) {
+			return false;
+		}
+
+		$value = $this->calculate_cogs_value_core();
+
+		/**
+		 * Filter to modify the Cost of Goods Sold value that gets calculated for a given order item.
+		 *
+		 * @since 9.5.0
+		 *
+		 * @param float|null $value The value originally calculated, null if it was not possible to calculate it.
+		 * @param WC_Order_Item $line_item The order item for which the value is calculated.
+		 */
+		$value = apply_filters( 'woocommerce_calculated_order_item_cogs_value', $value, $this );
+
+		if ( is_null( $value ) ) {
+			return false;
+		}
+
+		$this->set_cogs_value( (float) $value );
+		return true;
+	}
+
+	// phpcs:disable Squiz.Commenting.FunctionComment.InvalidNoReturn
+
+	/**
+	 * Core method to calculate the Cost of Goods Sold value for this line item:
+	 * it doesn't check if COGS is enabled at class or system level, doesn't fire hooks, and doesn't set the value as the current one for the line item.
+	 *
+	 * @return float|null The calculated value, or null if the value can't be calculated for some reason.
+	 * @throws Exception The class doesn't implement its own version of this method. Derived classes are expected to override this method when has_cogs returns true.
+	 */
+	protected function calculate_cogs_value_core(): ?float {
+		// phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+		throw new Exception(
+			sprintf(
+				// translators: %1$s = class and method name.
+				__( 'Method %1$s is not implemented. Classes overriding has_cogs must override this method too.', 'woocommerce' ),
```

Comment:
Why not escape (vs disabling the `ExceptionNotEscaped` rule)?

Reply:
This exception signals a "design time" error: if you get it it means that your code is wrong (you did override `has_cogs` to return true but didn't override `calculate_cogs_value_core`) and needs fixing. You are never going to get this in production, so it doesn't make much sense to add escaping.